### ISSUE TITLE
6.2.2/dtest-release: test_mv_populating_from_existing_data_during_node_remove fails with View ks.tm_table_mv_1 not built

### ISSUE THREAD

**Issue by mykaul** [scylladb/scylladb#21829]
Log snippet:
```
13:41:45,971 703     tools.tables_view_manager      DEBUG    tables_view_manager.py:579  | test_mv_populating_from_existing_data_during_node_remove: Materialized view tm_table_mv_9 has been created
13:41:46,627 703     materialized_views_test        DEBUG    materialized_views_test.py:385  | test_mv_populating_from_existing_data_during_node_remove: START: remove node node2
13:41:46,627 703     tools.misc                     DEBUG    misc.py             :328  | test_mv_populating_from_existing_data_during_node_remove: Stopping node node2 (hostid daf06cfe-1664-47c6-b5e7-6530733c6572) gently=False
13:42:08,251 703     tools.misc                     DEBUG    misc.py             :330  | test_mv_populating_from_existing_data_during_node_remove: Remove node node2 (hostid daf06cfe-1664-47c6-b5e7-6530733c6572)
13:42:08,438 703     ccm                            DEBUG    cluster.py          :754  | test_mv_populating_from_existing_data_during_node_remove: node1: nodetool cmd=['/jenkins/workspace/scylla-6.2/dtest-release/scylla/.dtest/dtest-3ezuf_p6/test/node1/bin/scylla', 'nodetool', '-h', '127.0.58.1', '-p', '10000', 'removenode', 'daf06cfe-1664-47c6-b5e7-6530733c6572'] wait=True timeout=None
13:42:18,918 703     cassandra.cluster              INFO     thread.py           :58   | test_mv_populating_from_existing_data_during_node_remove: Cassandra host 127.0.58.2:9042 removed
13:42:18,939 703     ccm                            DEBUG    cluster.py          :754  | test_mv_populating_from_existing_data_during_node_remove: node1: checking 
13:42:18,939 703     materialized_views_test        DEBUG    materialized_views_test.py:406  | test_mv_populating_from_existing_data_during_node_remove: FINISH: remove node node2
13:42:18,939 703     tools.data                     DEBUG    data.py             :259  | test_mv_populating_from_existing_data_during_node_remove: '2' threads finished at 2024-12-01 13:42:18.939921
13:42:18,940 703     tools.tables_view_manager      DEBUG    tables_view_manager.py:679  | test_mv_populating_from_existing_data_during_node_remove: Waiting for view ks.tm_table_mv_0 to finish building: num_nodes=3 num_alive=3 cl=QUORUM
13:42:18,943 703     tools.tables_view_manager      DEBUG    tables_view_manager.py:688  | test_mv_populating_from_existing_data_during_node_remove: wait_for_view ks.tm_table_mv_0: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('18fc44e9-40ba-4613-b8ab-bdf0e9cd2f96')}, 'STARTED': {UUID('2fd5324f-952a-4faf-a495-d48262d1894b'), UUID('aa6da97a-5ac6-4f1d-87f3-a3e94899ab36')}})
...
13:52:29,256 703     tools.tables_view_manager      DEBUG    tables_view_manager.py:693  | test_mv_populating_from_existing_data_during_node_remove: wait_for_view ks.tm_table_mv_1: node 18fc44e9-40ba-4613-b8ab-bdf0e9cd2f96/127.0.58.1 is not done yet
13:52:34,513 703     errors                         ERROR    conftest.py         :217  | test_mv_populating_from_existing_data_during_node_remove: test failed: 
self = <materialized_views_test.TestMaterializedViews object at 0x7fab3a565280>

    def test_mv_populating_from_existing_data_during_node_remove(self):
        """Create 10 materialized views in parallel with removing a node"""
>       self._mv_populating_from_existing_data_during_changes_test("remove node")

materialized_views_test.py:764: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
materialized_views_test.py:971: in _mv_populating_from_existing_data_during_changes_test
    wait_for_view(cluster=self.cluster, session=session, ks=tm.keyspace, view=mv_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cluster = <ccmlib.scylla_cluster.ScyllaCluster object at 0x7fab346e2420>
session = <cassandra.cluster.Session object at 0x7fab038c6d20>, ks = 'ks'
view = 'tm_table_mv_1', raise_exception = True, timeout = 600, cl = 4

    def wait_for_view(cluster, session, ks, view, raise_exception=True, timeout=None, cl=None):  # noqa: PLR0913
        num_nodes = len(cluster.nodelist())
        num_alive = len([node for node in cluster.nodelist() if node.is_live()])
        if cl is None:
            cl = ConsistencyLevel.ONE if num_alive == 1 else ConsistencyLevel.QUORUM
        cl_name = ConsistencyLevel.value_to_name[cl]
        logger.debug(f"Waiting for view {ks}.{view} to finish building: num_nodes={num_nodes} num_alive={num_alive} cl={cl_name}")
    
        query = SimpleStatement(view_built_status_query(ks, view, "host_id,status"), consistency_level=cl)
    
        def _view_build_finished_on_live_nodes():
            status = defaultdict(set)
            entries = rows_to_list(session.execute(query))
            for entry in entries:
                status[entry[1]].add(entry[0])
            logger.debug(f"wait_for_view {ks}.{view}: status={status}")
            done = status["SUCCESS"]
            for node in cluster.nodelist():
                try:
                    if node.is_live() and not (UUID(node.hostid()) in done):
                        logger.debug(f"wait_for_view {ks}.{view}: node {node.hostid()}/{node.address()} is not done yet")
                        return False
                except NodetoolError:
                    # If we decomissioned a node with "nodetool decommission"
                    # the code above may temporarily think that node.is_alive()
                    # is still true, but node.hostid(), which calls nodetool,
                    # can fail with an exception. In this case we just need to
                    # consider this node non-live.
                    pass
            return True
    
        if not timeout:
            timeout = 600
            if hasattr(cluster, "scylla_mode") and cluster.scylla_mode == "debug":
                timeout = 900
    
        deadline = time.time() + timeout
        while time.time() < deadline:
            if _view_build_finished_on_live_nodes():
                return True
            time.sleep(5)
    
        error_msg = f"View {ks}.{view} not built"
        if raise_exception:
>           raise Exception(error_msg)
E           Exception: View ks.tm_table_mv_1 not built

tools/tables_view_manager.py:717: Exception
13:52:34,513 703     errors                         ERROR    conftest.py         :217  | test_mv_populating_from_existing_data_during_node_remove: test failed: 
self = <materialized_views_test.TestMaterializedViews object at 0x7fab3a565280>

    def test_mv_populating_from_existing_data_during_node_remove(self):
        """Create 10 materialized views in parallel with removing a node"""
>       self._mv_populating_from_existing_data_during_changes_test("remove node")

materialized_views_test.py:764: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
materialized_views_test.py:971: in _mv_populating_from_existing_data_during_changes_test
    wait_for_view(cluster=self.cluster, session=session, ks=tm.keyspace, view=mv_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cluster = <ccmlib.scylla_cluster.ScyllaCluster object at 0x7fab346e2420>
session = <cassandra.cluster.Session object at 0x7fab038c6d20>, ks = 'ks'
view = 'tm_table_mv_1', raise_exception = True, timeout = 600, cl = 4

    def wait_for_view(cluster, session, ks, view, raise_exception=True, timeout=None, cl=None):  # noqa: PLR0913
        num_nodes = len(cluster.nodelist())
        num_alive = len([node for node in cluster.nodelist() if node.is_live()])
        if cl is None:
            cl = ConsistencyLevel.ONE if num_alive == 1 else ConsistencyLevel.QUORUM
        cl_name = ConsistencyLevel.value_to_name[cl]
        logger.debug(f"Waiting for view {ks}.{view} to finish building: num_nodes={num_nodes} num_alive={num_alive} cl={cl_name}")
    
        query = SimpleStatement(view_built_status_query(ks, view, "host_id,status"), consistency_level=cl)
    
        def _view_build_finished_on_live_nodes():
            status = defaultdict(set)
            entries = rows_to_list(session.execute(query))
            for entry in entries:
                status[entry[1]].add(entry[0])
            logger.debug(f"wait_for_view {ks}.{view}: status={status}")
            done = status["SUCCESS"]
            for node in cluster.nodelist():
                try:
                    if node.is_live() and not (UUID(node.hostid()) in done):
                        logger.debug(f"wait_for_view {ks}.{view}: node {node.hostid()}/{node.address()} is not done yet")
                        return False
                except NodetoolError:
                    # If we decomissioned a node with "nodetool decommission"
                    # the code above may temporarily think that node.is_alive()
                    # is still true, but node.hostid(), which calls nodetool,
                    # can fail with an exception. In this case we just need to
                    # consider this node non-live.
                    pass
            return True
    
        if not timeout:
            timeout = 600
            if hasattr(cluster, "scylla_mode") and cluster.scylla_mode == "debug":
                timeout = 900
    
        deadline = time.time() + timeout
        while time.time() < deadline:
            if _view_build_finished_on_live_nodes():
                return True
            time.sleep(5)
    
        error_msg = f"View {ks}.{view} not built"
        if raise_exception:
>           raise Exception(error_msg)
E           Exception: View ks.tm_table_mv_1 not built

tools/tables_view_manager.py:717: Exception
```

node 2 is gone, but I do see in node 3, 4 that view building is complete, but node 1 indeed fails building the view.
[dtest-gw7.log](https://github.com/user-attachments/files/18052019/dtest-gw7.log)
[node4.log](https://github.com/user-attachments/files/18052017/node4.log)
[node3.log](https://github.com/user-attachments/files/18052016/node3.log)
[node1.log](https://github.com/user-attachments/files/18052018/node1.log)

**Comment by mykaul** [comment]
note - it was streamed to node 1, perhaps that can partially explain this:
```
INFO  2024-12-01 13:42:10,880 [shard 0:strm] stream_session - [Stream #10adac00-afea-11ef-9a73-adae88c8d919] Executing streaming plan for Tablet rebuild-ks-index-0 with peers=127.0.58.4, slave
INFO  2024-12-01 13:42:10,881 [shard 0:strm] stream_session - [Stream #10adac00-afea-11ef-9a73-adae88c8d919] Skip sending ks=ks, cf=tm_table_mv_1, reader contains no data, with new rpc streaming
INFO  2024-12-01 13:42:10,882 [shard 1:strm] stream_session - [Stream #10adac00-afea-11ef-9a73-adae88c8d919] Start sending ks=ks, cf=tm_table_mv_1, estimated_partitions=640, with new rpc streaming
INFO  2024-12-01 13:42:10,889 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {daf06cfe-1664-47c6-b5e7-6530733c6572}
INFO  2024-12-01 13:42:10,893 [shard 0:strm] stream_session - [Stream #10adac00-afea-11ef-9a73-adae88c8d919] Streaming plan for Tablet rebuild-ks-index-0 succeeded, peers={127.0.58.4}, tx=90 KiB, 7415.11 KiB/s, rx=0 KiB, 0.00 KiB/s
...
INFO  2024-12-01 13:42:11,271 [shard 1:strm] stream_session - [Stream #10e38910-afea-11ef-90fb-e7077881ff5a] Write to sstable for ks=ks, cf=tm_table_mv_1, estimated_partitions=640, received_partitions=567
INFO  2024-12-01 13:42:11,273 [shard 0:strm] stream_session - [Stream #10e38910-afea-11ef-90fb-e7077881ff5a] Streaming plan for Tablet rebuild-ks-index-0 succeeded, peers={127.0.58.3}, tx=0 KiB, 0.00 KiB/s, rx=94 KiB, 2440.43 KiB/s
...
INFO  2024-12-01 13:42:11,312 [shard 1:strm] stream_session - [Stream #10e38911-afea-11ef-90fb-e7077881ff5a] Write to sstable for ks=ks, cf=tm_table_mv_1, estimated_partitions=0, received_partitions=10042
INFO  2024-12-01 13:42:11,313 [shard 0:strm] stream_session - [Stream #10e38911-afea-11ef-90fb-e7077881ff5a] Streaming plan for Tablet rebuild-ks-index-0 succeeded, peers={127.0.58.4}, tx=0 KiB, 0.00 KiB/s, rx=1668 KiB, 20960.61 KiB/s
INFO  2024-12-01 13:42:11,313 [shard 0:strm] range_streamer - Finished 2 out of 2 ranges for rebuild, finished percentage=1
INFO  2024-12-01 13:42:11,313 [shard 0:strm] range_streamer - Tablet rebuild with 127.0.58.4 for keyspace=ks succeeded, took 0.07967563 seconds
INFO  2024-12-01 13:42:11,313 [shard 0:strm] range_streamer - Tablet rebuild succeeded, took 0 seconds, nr_ranges_remaining=0
INFO  2024-12-01 13:42:11,314 [shard 0: gms] raft_topology - updating topology state: advance fence version to 43
```

**Comment by Annamikhlin** [comment]
the same issue seen on master/dtest-release - 
https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/785/testReport/junit/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split001___test_mv_populating_from_existing_data_during_node_decommission/

**Comment by mlitvk** [comment]
I suspect the view builder enters a loop where it builds the same token ranges repeatedly and it doesn't detect when it should exit correctly.

We check here in the view builder consumer whether a view is built:
```cpp
    void check_for_built_views() {                                                                                    
        inject_failure("view_builder_check_for_built_views");                                                         
        for (auto it = _step.build_status.begin(); it != _step.build_status.end();) {                                 
            // A view starts being built at token t1. Due to resharding, that may not necessarily be a                
            // shard-owned token. We finish building the view when the next_token to build is just before             
            // (or at) the first token, but the shard-owned current token is after (or at) the first token.           
            // In the system tables, we set first_token = next_token to signal the completion of the build            
            // process in case of a restart.                                                                          
            if (it->next_token && *it->next_token <= it->first_token && _step.current_token() >= it->first_token) {   
                _built_views.views.push_back(std::move(*it));                                                         
                it = _step.build_status.erase(it);                                                                    
            } else {                                                                                                  
                ++it;                                                                                                 
            }                                                                                                         
        }                                                                                                             
    }                                                                                                                 
```

I added a print in this function. here is a visualization of `it->first_token` (F) and `_step.current_token()` (C) in a reproduction of the issue for a specific shard and view:

https://github.com/user-attachments/assets/8a9fca1d-11df-4ff3-a2ec-eec64b88eda3

the view building is repeated this way for 10 minutes.

logs from the wraparound:
```
INFO  2025-01-22 12:57:20,816 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(-630325686913624) first 53309345463726304 current -508538429469205 built=false
INFO  2025-01-22 12:57:20,816 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(-508538429469205) first 53309345463726304 current -446984407858504 built=false
INFO  2025-01-22 12:57:20,816 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(-446984407858504) first 53309345463726304 current -427370705545089 built=false
INFO  2025-01-22 12:57:20,817 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(-427370705545089) first 53309345463726304 current -218476145641997 built=false
INFO  2025-01-22 12:57:20,817 [shard 1: gms] view - Completed build step for base ks.tm_table, at token -218476145641997; views=["tm_table_mv_9", "tm_table_mv_8", "tm_table_mv_7", "tm_table_mv_6", "tm_table_mv_5", "tm_table_mv_4", "tm_table_mv_3", "tm_table_mv_2", "tm_table_mv_1"]
INFO  2025-01-22 12:57:20,817 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(minimum token) first 53309345463726304 current minimum token built=false
INFO  2025-01-22 12:57:20,817 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(minimum token) first 53309345463726304 current -9223317535497729914 built=false
INFO  2025-01-22 12:57:20,818 [shard 1: gms] view - MICHAEL check tm_table_mv_3: next optional(-9223317535497729914) first 53309345463726304 current -9222784767296949215 built=false
```

it seems the condition `_step.current_token() >= it->first_token` is never satisfied.

we update the current_token here when consuming a new partition and then check for built views:
```
    stop_iteration consume_new_partition(const dht::decorated_key& dk) {                                              
        inject_failure("view_builder_consume_new_partition");                                                         
        if (dk.key().is_empty()) {                                                                                    
            on_internal_error(vlogger, format("Trying to consume empty partition key {}", dk));                       
        }                                                                                                             
        _step.current_key = std::move(dk);                                                                            
        check_for_built_views();                                                                                      
```

then in `consume_end_of_stream` we go back to the minimum token.

Perhaps when we started view building we had a partition with the token `first_token`, but after a range movement (when a node is removed) we don't have a partition with a token `>= first_token` anymore?

**Comment by mlitvk** [comment]
reproducer:
```python
@pytest.mark.asyncio
async def test_mv_build_with_tablet_move(manager: ManagerClient):
    servers = [await manager.server_add()]

    await manager.api.disable_tablet_balancing(servers[0].ip_addr)

    cql = manager.get_cql()
    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} AND tablets = {'initial': 16}")
    await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int)")

    # prefill the base table with enough rows so that view building takes some time
    # and runs during the tablet move
    keys = 500000
    concurrency = 1000
    for k in range(0, keys, concurrency):
        await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({i}, {i})") for i in range(k, k+concurrency)])

    logger.info("Adding new server")
    servers.append(await manager.server_add())

    # create some views so they are built together but starting at different tokens
    view_count = 10
    for i in range(view_count):
        await cql.run_async(f"CREATE MATERIALIZED VIEW test.tv{i} AS SELECT * FROM test.test WHERE c IS NOT NULL AND pk IS NOT NULL PRIMARY KEY (c, pk)")
        await asyncio.sleep(1)

    s0_host_id = await manager.get_host_id(servers[0].server_id)
    s1_host_id = await manager.get_host_id(servers[1].server_id)
    dst_shard = 0

    # move all tablets except the first one (with lowest token range) to the other node.
    table_id = await manager.get_table_id('test', 'test')
    rows = await manager.cql.run_async(f"SELECT last_token FROM system.tablets where table_id = {table_id}")
    move_tablets_tasks = []
    for r in rows[1:]:
        tablet_token = r.last_token
        replica = await get_tablet_replica(manager, servers[0], 'test', 'test', tablet_token)
        logger.info(f"move tablet token={tablet_token}")
        move_tablets_tasks.append(asyncio.create_task(manager.api.move_tablet(servers[0].ip_addr, "test", "test", replica[0], replica[1], s1_host_id, dst_shard, tablet_token)))
    await asyncio.gather(*move_tablets_tasks)

    for i in range(view_count):
        await wait_for_view(cql, f"tv{i}", len(servers))

```

1. start with 16 tablets of the base table in the first node
2. create few views that start to build at different tokens:
```
INFO  2025-01-23 12:45:32,923 [shard 0: gms] view - Building view test.tv0, starting at token minimum token
INFO  2025-01-23 12:45:34,477 [shard 0: gms] view - Building view test.tv1, starting at token -8664177871504282604
INFO  2025-01-23 12:45:36,510 [shard 0: gms] view - Building view test.tv2, starting at token -6684816878496489145
INFO  2025-01-23 12:45:38,362 [shard 0: gms] view - Building view test.tv3, starting at token -6244205401827952674
INFO  2025-01-23 12:45:39,565 [shard 0: gms] view - Building view test.tv4, starting at token -5995081919536463432
INFO  2025-01-23 12:45:41,392 [shard 0: gms] view - Building view test.tv5, starting at token -4403386516381294504
INFO  2025-01-23 12:45:42,618 [shard 0: gms] view - Building view test.tv6, starting at token -4176837721588416989
INFO  2025-01-23 12:45:44,461 [shard 0: gms] view - Building view test.tv7, starting at token -3836740547251052230
INFO  2025-01-23 12:45:46,489 [shard 0: gms] view - Building view test.tv8, starting at token -3504326481207717646
INFO  2025-01-23 12:45:48,541 [shard 0: gms] view - Building view test.tv9, starting at token -2069251178505130784
```
3. move all tablets except the first one to the other node
```
12:45:50.350 INFO>  move tablet token=-6917529027641081857
12:45:50.357 INFO>  move tablet token=-5764607523034234881
12:45:50.373 INFO>  move tablet token=-4611686018427387905
12:45:50.393 INFO>  move tablet token=-3458764513820540929
12:45:50.411 INFO>  move tablet token=-2305843009213693953
12:45:50.428 INFO>  move tablet token=-1152921504606846977
12:45:50.450 INFO>  move tablet token=-1
12:45:50.467 INFO>  move tablet token=1152921504606846975
12:45:50.486 INFO>  move tablet token=2305843009213693951
12:45:50.511 INFO>  move tablet token=3458764513820540927
12:45:50.528 INFO>  move tablet token=4611686018427387903
12:45:50.556 INFO>  move tablet token=5764607523034234879
12:45:50.569 INFO>  move tablet token=6917529027641081855
12:45:50.587 INFO>  move tablet token=8070450532247928831
12:45:50.604 INFO>  move tablet token=9223372036854775807
```
4. the first two views that started to build at low tokens finish building
```
INFO  2025-01-23 12:45:58,444 [shard 0:strm] view - Finished building view test.tv0
INFO  2025-01-23 12:46:00,883 [shard 0:strm] view - Finished building view test.tv1
```
5. the building of the other views is stuck in a loop for additional 2 minutes until timeout
```
INFO  2025-01-23 12:48:03,069 [shard 0:strm] view - Completed build step for base test.test, at token -8172139367854932691; views=["tv9", "tv8", "tv7", "tv6", "tv5", "tv4", "tv3", "tv2"]
INFO  2025-01-23 12:48:03,083 [shard 0:strm] view - Completed build step for base test.test, at token -8166932885961843813; views=["tv9", "tv8", "tv7", "tv6", "tv5", "tv4", "tv3", "tv2"]
INFO  2025-01-23 12:48:03,103 [shard 0:strm] view - Completed build step for base test.test, at token -8163240673057160431; views=["tv9", "tv8", "tv7", "tv6", "tv5", "tv4", "tv3", "tv2"]
```

**Comment by mlitvk** [comment]
After fixing the issue decribed above the test may still fail on another issue of view inconsistency that happens when we move tablets during view building:
```
FAILED materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_node_remove[957-1024] - AssertionError: Expected a row count of 40000 in table 'tm_table_mv_0', but got 38915
==================================== 1 failed, 1023 passed in 12248.05s (3:24:08) ====================================
```
This appears to be a known issue: https://github.com/scylladb/scylladb/issues/18826

**Comment by lsfreitas** [comment]
Seems like the same issue here: https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/806/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split001___test_mv_populating_from_existing_data_during_node_decommission/

[1738020955048_materialized_views_test.py__TestMaterializedViews__test_mv_populating_from_existing_data_during_node_decommission.zip](https://github.com/user-attachments/files/18573771/1738020955048_materialized_views_test.py__TestMaterializedViews__test_mv_populating_from_existing_data_during_node_decommission.zip)
